crates/sui-types/src/move_package.rs

```diff
@@ -94,11 +104,22 @@ impl MovePackage {
         version: SequenceNumber,
         module_map: &BTreeMap<String, Vec<u8>>,
         max_move_package_size: u64,
+    ) -> Result<Self, ExecutionError> {
+        Self::new_with_type_origin(id, version, module_map, max_move_package_size, None)
```

Comment:
In terms of how to structure initialisation for `MovePackage` thoughts on the following structure (+ equivalents that accepts an iterator of `CompiledModule`).

- `MovePackage::new` which is responsible for creating a brand new package (i.e. newly published) which doesn't accept a version (it will be `OBJECT_START_VERSION`), and calculates its type origin table based on its own modules.
- `MovePackage::upgrade` which takes a reference to an existing `MovePackage`, as well as the new modules, an `id`, and a `max_move_package_size`.  This creates a package where:
   - The ID is one greater than the ID of the referenced package.
   - The type origin table is derived from the predecessors, and augmented with entries from the new modules.
   - When the linkage table exists, we can do the versioning check here.
   - We could also potentially run the compatibility check here? cc @tzakian.

Reply:
Love the idea of both `MovePackage::new` and `MovePackage::upgrade`; the more structure we can have (and the least amount of guesswork at the callsite) the better IMO. I'm happy to add and implement the `upgrade` over in #9122 if we want to :) I'm guessing the implementation for `new` will probably want to be in this PR?